research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -326,43 +322,60 @@ def test(datasets, net, epoch, writer=None, model_directory=None,
     qvecs = qvecs.numpy()
 
     # Search, rank and print test set metrics.
-    scores = np.dot(vecs.T, qvecs)
-    ranks = np.transpose(np.argsort(-scores, axis=0))
-    metrics = global_features_utils.compute_metrics_and_print(dataset, ranks,
-                                                              cfg['gnd'])
-    # Save calculated metrics in a tensorboard format.
-    if writer:
-      tf.summary.scalar('test_accuracy_{}_E'.format(dataset), metrics[0][0],
-                        step=epoch)
-      tf.summary.scalar('test_accuracy_{}_M'.format(dataset), metrics[1][0],
-                        step=epoch)
-      tf.summary.scalar('test_accuracy_{}_H'.format(dataset), metrics[2][0],
-                        step=epoch)
-      writer.flush()
+    _calculate_metrics_and_export_to_tensorboard(vecs, qvecs, dataset, cfg,
+                                                 writer, epoch, whiten=False)
 
     if learned_whitening is not None:
-
       # Whiten the vectors.
       mean_vector = learned_whitening['m']
       projection_matrix = learned_whitening['P']
       vecs_lw = whiten.whitenapply(vecs, mean_vector, projection_matrix)
       qvecs_lw = whiten.whitenapply(qvecs, mean_vector, projection_matrix)
 
       # Search, rank, and print.
-      scores = np.dot(vecs_lw.T, qvecs_lw)
-      ranks = np.transpose(np.argsort(-scores, axis=0))
-      metrics = global_features_utils.compute_metrics_and_print(
-              dataset + ' + whiten', ranks, cfg['gnd'])
-
-      if writer:
-        tf.summary.scalar('test_accuracy_whiten_{}_E'.format(dataset),
-                          metrics[0][0], step=epoch)
-        tf.summary.scalar('test_accuracy_whiten_{}_M'.format(dataset),
-                          metrics[1][0], step=epoch)
-        tf.summary.scalar('test_accuracy_whiten_{}_H'.format(dataset),
-                          metrics[2][0], step=epoch)
-        writer.flush()
+      _calculate_metrics_and_export_to_tensorboard(
+              vecs_lw, qvecs_lw, dataset, cfg, writer, epoch, whiten=True)
```

Comment:
hmmm, do you need to call this with `dataset + ' + whiten'` instead of simply `dataset`?
Or, as you have done below, add the `' + whiten'` in the new function below, depending on the value of the `whiten` argument.

Reply:
It should be working without " + whiten". The " + whiten" part in the previous code iteration was used for the naming of the tensorboard variable. Now, we overcome the need of having it by the `whiten=True` in the function, which results in choosing suitable metric names (i.e., 'test_accuracy_whiten_{}_E' instead of just 'test_accuracy_{}_E').

Reply:
Right, but I was talking about the line `metrics = global_features_utils.compute_metrics_and_print(dataset, ranks,
                                                            cfg['gnd'])`

Before, you had `dataset + ' + whiten'` instead of `dataset` in the above line, when you were logging the whitening case.

Anyway, it's not a big issue at all, it will I guess just log things slightly differently. Feel free to ignore this, or change the code to follow the same behavior as before -- as you prefer. Feel free to resolve this comment as well.